doc/dev/msgr2.rst

```diff
@@ -108,14 +108,18 @@ Authentication
 * TAG_AUTH_REQUEST: client->server::
 
     __le32 method;  // CEPH_AUTH_{NONE, CEPHX, ...}
+    __le32 num_preferred_modes;
+    list<__le32> mode  // CEPH_CON_MODE_*
```

Comment:
I'm divided here.  On one hand, all things being equal, I want the server's preferences to prevail.  But I also think that if the server says 'crc,secure' (meaning we usually use crc but client coudl force secure if they want) and the client says 'secure crc' (client prefers secure) then we probably should do secure?  Or maybe not... if the client really wants secure it can specify only secure and the right thing will happen.

So I think it's either

- what we have now, where server config is in preference order and client order doesn't matter, or
- we rank the modes so that if either client or server prefers a 'greater' mode (secure > crc?) then we use it, and so both client and server configs are preference order.

Reply:
I think we should assume that any real preference the server has will be about whether a mode is allowed at all, not about the order they go in. That’s what we do for the auth method — client asks for a particular method, the server accepts it or denies it